src/lib.rs

```diff
@@ -1,215 +1,541 @@
+//! Astarte is an Open Source IoT platform focused on Data management.
+//! It takes care of everything from collecting data from devices to delivering data to end-user applications.
+//! To achieve such a thing, it uses a mixture of mechanisms and paradigm to store organized data, perform live queries.
+//!
+//! <https://docs.astarte-platform.org/>
+
 mod crypto;
 mod interface;
 mod pairing;
+pub mod types;
 
+use bson::{to_document, Bson};
 use crypto::Bundle;
+use itertools::Itertools;
+use log::{debug, trace};
 use openssl::error::ErrorStack;
 use pairing::PairingError;
-use rumqttc::{AsyncClient, ClientConfig, Event, MqttOptions};
+use rumqttc::EventLoop;
+use rumqttc::{AsyncClient, ClientConfig, Event, MqttOptions, Transport};
 use rustls::{internal::pemfile, Certificate, PrivateKey};
 use std::collections::HashMap;
 use std::fmt;
 use std::path::Path;
 use std::sync::Arc;
-use tokio::task::JoinHandle;
+use types::AstarteType;
 use url::Url;
 
 use interface::traits::Interface as InterfaceTrait;
 pub use interface::Interface;
 
-pub struct Device {
+use crate::interface::Ownership;
+
+/// Astarte client
+#[derive(Clone)]
+pub struct AstarteSdk {
     realm: String,
     device_id: String,
     credentials_secret: String,
     pairing_url: String,
     private_key: PrivateKey,
     csr: String,
-    certificate_pem: Option<Vec<Certificate>>,
-    broker_url: Option<Url>,
-    client: Option<AsyncClient>,
-    eventloop_task: Option<JoinHandle<()>>,
+    certificate_pem: Vec<Certificate>,
+    broker_url: Url,
+    client: AsyncClient,
+    eventloop: Arc<tokio::sync::Mutex<EventLoop>>,
     interfaces: HashMap<String, Interface>,
 }
 
-pub struct DeviceBuilder {
+/// Builder for Astarte client
+///
+/// ```
+/// use astarte_sdk::AstarteOptions;
+///
+/// let realm = "test";
+/// let device_id = "2TBn-jNESuuHamE2Zo1anA";
+/// let credentials_secret = "G6PjaFRwhjKLb07hk/lAUHqFoI+aIg39CEQ1MQ+mdZk=";
+/// let pairing_url = "https://api.eu1.astarte.cloud/pairing";
```

Comment:
Remove explicit Astarte info from comments

Reply:
You mean the pairing url? The rest of the variables are all fake data

Reply:
You mean the pairing url? The rest of the variables are all fake data